File: 6.迭代器模式/geqiangbao_20201125_iterator_OC/geqiangbao_20201125_iterator_OC/demo/SetIterator.m
//
//  SetIterator.m
//  geqiangbao_20201125_iterator_OC
//
//  Created by xiaobao on 2020/11/25.
//

#import "SetIterator.h"

@interface SetIterator ()
@property (nonatomic,strong) NSEnumerator<id> * iterator;
@end

@implementation SetIterator

-(instancetype)initWith:(id)datas{
    self = [super init];
    if (self) {
        self.datas = datas;
        _iterator = self.datas.objectEnumerator;
    }
    return self;
}

- (id)next{
    if ([self hasNext]) {
        self.index++;
        return self.iterator.nextObject;
    }
    return nil;
   
}

- (BOOL)hasNext{
    return self.index < self.datas.count;
}

@end

File: 6.迭代器模式/geqiangbao_20201125_iterator_swift/geqiangbao_20201125_iterator_swift/demo/SetIterator.swift
//
//  SetIterator.swift
//  geqiangbao_20201125_iterator_swift
//
//  Created by xiaobao on 2020/11/25.
//

import UIKit

class SetIterator<T>: AnyIterator<NSSet,T> {

    var iterator:NSFastEnumerationIterator?
    override init(datas: NSSet) {
        super.init(datas: datas)
        self.iterator = self.datas?.makeIterator()
    }
    
    override func hasNext() -> Bool {
        return self.index < self.datas?.count ?? 0
    }
    
    override func next() -> T? {
        if self.hasNext() {
            self.index += 1
            let a = self.iterator?.next() as? T
            return a
        }
        return nil
    }
}
